### PR TITLE
Improve controller connection handling.

### PR DESCRIPTION
**Pull request by frankban**
Track anonymous mode so that it is not only based on URL state, but potentially kept enabled regardless of user navigation. 
Also start the controller pinger as soon as the controller is connected, without waiting for the user to be logged in. This should prevent disconnections due to timeouts when in anonymous mode (in which the controller is connected but not yet authenticated).

Also increased the mocha "done" timeout from 2 to 5 seconds (while waiting for a new laptop).

Fixes https://github.com/juju/juju-gui/issues/2509

### CODE REVIEW

**Comment by jujugui** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://ci-cge.jujugui.org:8080//job/juju-gui/6842/

**Comment by jujugui** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://ci-cge.jujugui.org:8080//job/juju-gui/6843/

**Comment by frankban** [comment]
**QA** must be done in both gijoe and gisf scenarios.
For gisf, at this stage you can use `guiproxy -config 'gisf: true, jujuEnvUUID: ""'`.

In **gijoe**, we only have to check that logging in and out works as usual, with both userpass and macaroons. You should nevere end up in anonymous mode here.

In **gisf**, automatic macaroon auth should be performed when logged out and navigating all URLs except /new. When going to /new, anonymous mode should be enabled. This means that you can work on the canvas as an anonymous user. When in anonymous mode, add a charm to the canvas, click deploy and the deployment flow should open up and give you the option to log in. DO NOT LOGIN yet, just wait on that page for a couple of minutes, and ensure nothing happens (basically you should not be forced to log in in any way, see the linked bug). Now, to simulate a proper disconnection, stop the guiproxy for a couple of seconds, and the restart it. Ensure nothing happened in the GUI, that you are still at the beginning of the deployment flow. Now proceed with the authentication process and deploy the charm. When done, ensure `app.anonymousMode` is false (using the JS console), and that the authenticated interaction with the GUI works as usual.

File: jujugui/static/gui/src/app/app.js [hunk 1]
```diff
@@ -407,6 +407,9 @@ YUI.add('juju-gui', function(Y) {
       */
       this.userPaths = new Map();
 
+      // Track anonymous mode.
+      this.anonymousMode = false;
```
Comment (Lukewh) [inline jujugui/static/gui/src/app/app.js]: Is this the default state? Do we want it to be true by default? Whichever way, maybe add a little comment as to why it's one or the other?

Comment (frankban) [inline jujugui/static/gui/src/app/app.js]: Done.

**Comment by jujugui** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://ci-cge.jujugui.org:8080//job/juju-gui/6844/

**Comment by jaycee** [comment]
QA (gisf) notes:

Following the provided instructions using staging, everything worked as described.

When I tried it on prod, I got an error back from IdM that confuses me (I don't think it's related to this branch, but just in case it seems worth mentioning).

2017-02-23-081548_1267x234_scrot

GISF QA OK. Trying GIJoe now.

**Comment by jujugui** [comment]
Refer to this link for build results (access rights to CI server needed): 
http://ci-cge.jujugui.org:8080//job/juju-gui/6845/

**Comment by jaycee** [comment]
QA (gijoe) notes:

I spun up a controller with lxd using staging IdM as an identity provider. Log in worked both with controller admin creds and with usso creds (once I granted permission).

Interestingly, while login failed before I granted permission (as one might expect) I was still dumped to a canvas; we seem to be keying off of successful USSO login rather than successful connection to the controller on when to remove the login screen and go to canvas. This is not a blocker for this branch.

GIJOE QA OK.

File: jujugui/static/gui/src/app/app.js [hunk 2]
```diff
@@ -832,10 +838,15 @@ YUI.add('juju-gui', function(Y) {
...
         // If an anonymous GISF user lands on the GUI at /new then don't
         // attempt to log into the controller.
-        if (!creds.areAvailable && gisf &&
-            (currentState && currentState.root === 'new')) {
+        if ((
+          !creds.areAvailable && gisf && rootState === 'new'
+        ) || (
+          this.anonymousMode && rootState !== 'login'
+        )) {
+          this.anonymousMode = true;
```
Comment (jaycee) [inline jujugui/static/gui/src/app/app.js]: Assuming lint doesn't complain, I might indent this block one degree more--right now it lines up with the conditionals which took me a moment to parse out.

File: jujugui/static/gui/src/test/test_app.js [hunk 3]
```diff
@@ -2431,6 +2431,195 @@ describe('App', function() {
...
+      app.state = {current: {root: null}};
+      app._renderLogin = sinon.stub();
+      app.controllerAPI.after('login', evt => {
+        assert.strictEqual(app.anonymousMode, false, 'anonymousMode');
+        assert.strictEqual(app._renderLogin.called, true, '_renderLogin');
+        done();
+      });
+      app.controllerAPI.fire('login', {err: 'bad wolf'});
+    });
+  });
```
Comment (jaycee) [inline jujugui/static/gui/src/test/test_app.js]: Nice tests, thank you.

**Comment by frankban** [comment]
Thanks for the reviews!
:shipit:

**Comment by jujugui** [comment]
Status: merge request accepted. Url: http://ci-gce.jujugui.org:8080/job/juju-gui-merge

**Pull request merged**